### PR TITLE
Share Project Popover

### LINKED ISSUE #5815: Improve share project view
https://www.figma.com/design/Qt6EyotCBS3V6O31jVhMQ7/RILL-Latest?node-id=13326-670655&node-type=frame&t=4wfouJ2vanY74LBF-0

### PR DESCRIPTION
**Pull request by lovincyrus**
closes #5815

https://www.figma.com/design/Qt6EyotCBS3V6O31jVhMQ7/RILL-Latest?node-id=6801-232076&node-type=canvas&t=RbLXuShzLAEx5sVX-0

- [x] revisit list item name change
- [x] pending photo url in the avatar https://github.com/rilldata/rill/issues/5889 https://github.com/rilldata/rill/pull/5888
- [x] "You should be able to modify the roles for `all-users` – if not, let me know and we’ll fix it"
- [x] "Everyone at Rill Data"
- [x] add info icon for project perm manage members https://github.com/rilldata/rill/pull/5887#issuecomment-2415065713
- [x] https://github.com/rilldata/rill/pull/5887#discussion_r1799955918

CleanShot 2024-10-11 at 12 02 19@2x

### CODE REVIEW

**Review by ericpgreen2 — CHANGES_REQUESTED** [review]
UXQA:
1. Project Viewers should not be allowed to change a Project Member's role. We should draw from the `manage_project_members` project permission. 
    image
2. I don't see the ability to set the default role for "epg" Organization Members. The ability to set this role should draw from the `manage_org_members` organization permission.
    image
3. Should we hide the default "all-users" group from this UI? It represents all users in the organization, which is duplicative to the "Organization" section/setting. (CC @ericokuma, @jkhwu)
4. Should I be able to downgrade myself from "Admin" to "Viewer"? (CC @ericokuma, @jkhwu)
5. I'm seeing a `cursor-pointer` when I hover over the Organization and Group list items, yet nothing happens when I click.
6. I'm not seeing the text "Everyone from" [org] or "Everyone from" [group]. This could relate to this warning in my browser console:
    image

**Comment by lovincyrus** [comment]
> * Project Viewers should not be allowed to change a Project Member's role. We should draw from the `manage_project_members` project permission.
>   image

Only users that can manageProjectMembers will see the user invite button, see: https://github.com/rilldata/rill/blob/main/web-admin/src/features/navigation/TopNavigationBar.svelte#L180-L182

**Comment by ericokuma** [comment]
> UXQA:
> 
> * [x]  Project Viewers should not be allowed to change a Project Member's role. We should draw from the `manage_project_members` project permission.
>   image
> * [ ]  I don't see the ability to set the default role for "epg" Organization Members. The ability to set this role should draw from the `manage_org_members` organization permission.
> * [ ]  Should we hide the default "all-users" group from this UI? It represents all users in the organization, which is duplicative to the "Organization" section/setting. (CC @ericokuma, @jkhwu)
> * [x]  Should I be able to downgrade myself from "Admin" to "Viewer"? (CC @ericokuma, @jkhwu)
> * [x]  I'm seeing a `cursor-pointer` when I hover over the Organization and Group list items, yet nothing happens when I click.
> * [x]  I'm not seeing the text "Everyone from" [org] or "Everyone from" [group]. This could relate to this warning in my browser console:
>   image

1. Yes, we should hide the all-users group
2. Hmm, currently in other systems, you can downgrade yourself. You can also downgrade your role via CLI. So you should be able to downgrade yourself via UI?

**Comment by lovincyrus** [comment]
> * I don't see the ability to set the default role for "epg" Organization Members. The ability to set this role should draw from the `manage_org_members` organization permission.

I didn't implement this right off the bat because I wanted to keep this PR contained to project user management changeset. If we choose to include this org user role change from this project popover, we would have to individually set the role for each org user using `createAdminServiceSetOrganizationMemberUserRole`

```ts
{
  organization: string;
  email: string;
  data: {
    role?: string;
  };
}
```

and if we have 300 users in this project, we'd have to fire the request 300 times.

**Comment by ericpgreen2** [comment]
> Only users that can manageProjectMembers will see the user invite button, see: https://github.com/rilldata/rill/blob/main/web-admin/src/features/navigation/TopNavigationBar.svelte#L180-L182

Oh, I see what happened. My user was an Org Admin, so I got the `manage_project_members` permission [by default](https://github.com/rilldata/rill/blob/782bcf8ef51d5f1948517772afe93794256a0ea2/admin/permissions.go#L105-L127). 

So this is a case where I'd expect to see the exclamation mark w/ tooltip to indicate that the Project Viewer role is overridden/overpowered by the Organization Admin role.

image

**Comment by ericpgreen2** [comment]
> > * I don't see the ability to set the default role for "epg" Organization Members. The ability to set this role should draw from the `manage_org_members` organization permission.
> 
> I didn't implement this right off the bat because I wanted to keep this PR contained to project user management changeset. If we choose to include this org user role change from this project popover, we would have to individually set the role for each org user using `createAdminServiceSetOrganizationMemberUserRole`
> 
> ```ts
> {
>   organization: string;
>   email: string;
>   data: {
>     role?: string;
>   };
> }
> ```
> 
> and if we have [300 users](https://www.figma.com/design/Qt6EyotCBS3V6O31jVhMQ7/RILL-Latest?node-id=13326-670655&node-type=frame&t=cF5PVA45aU9jnfNk-0) in this project, we'd have to fire the request 300 times.

Oh I see. I agree that e.g. 300 individual requests would not be the way forward. We shouldn't be setting `user_orgs_roles` from this Project popover.

The designs seem to imply that we have an `organizations_projects_roles` table. We don't, but we do have a `usergroups_projects_roles` table. I could imagine an automatic `<organization>-users` usergroup for every organization, or I could imagine a dedicated `organizations_projects_roles` table. @begelundmuller @ericokuma, thoughts on this?

File: web-admin/src/features/projects/user-invite/UserInviteOrganization.svelte [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            fontSize="text-[10px]"
+            alt={user.userName}
+            bgColor={getRandomBgColor(user.userEmail)}
+          />
+          <li>{user.userName}</li>
+        </div>
+      {/each}
+      {#if memberUsers.length > 6}
+        <li>and {memberUsers.length - 6} more</li>
+      {/if}
```
[the remark this answers is not in the document]
Comment (ericpgreen2) [reply web-admin/src/features/projects/user-invite/UserInviteOrganization.svelte:50]: Which is the "project users table" you're referring to? In the mocks, I see users tables for Orgs & for User Groups, but not for Projects. But maybe I'm missing it or we have different terminology :)

When implemented, could we add a link from the Share Project Popover to the respective User tables, so someone can see the full list?

**Comment by ericpgreen2** [comment]
> Hmm, currently in other systems, you can downgrade yourself. You can also downgrade your role via CLI. So you should be able to downgrade yourself via UI?

@ericokuma, the "Share" button is only available to Project Admins, right? So if I use the popover to downgrade myself from Admin to Viewer, I'll be ejected from the popover I think. Open to anything, but thoughts on this?

**Comment by ericokuma** [comment]
> > > * I don't see the ability to set the default role for "epg" Organization Members. The ability to set this role should draw from the `manage_org_members` organization permission.
> > 
> > 
> > I didn't implement this right off the bat because I wanted to keep this PR contained to project user management changeset. If we choose to include this org user role change from this project popover, we would have to individually set the role for each org user using `createAdminServiceSetOrganizationMemberUserRole`
> > ```ts
> > {
> >   organization: string;
> >   email: string;
> >   data: {
> >     role?: string;
> >   };
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and if we have [300 users](https://www.figma.com/design/Qt6EyotCBS3V6O31jVhMQ7/RILL-Latest?node-id=13326-670655&node-type=frame&t=cF5PVA45aU9jnfNk-0) in this project, we'd have to fire the request 300 times.
> 
> Oh I see. I agree that e.g. 300 individual requests would not be the way forward. We shouldn't be setting `user_orgs_roles` from this Project popover.
> 
> The designs seem to imply that we have an `organizations_projects_roles` table. We don't, but we do have a `usergroups_projects_roles` table. I could imagine an automatic `<organization>-users` usergroup for every organization, or I could imagine a dedicated `organizations_projects_roles` table. @begelundmuller @ericokuma, thoughts on this?

Ah, the designs were just assuming that all users who have access to an org will have access to projects. I think have a dedicated organizations_projects_roles table might be nice although wouldn't the all-users user group suffice?

**Comment by ericokuma** [comment]
> > Hmm, currently in other systems, you can downgrade yourself. You can also downgrade your role via CLI. So you should be able to downgrade yourself via UI?
> 
> @ericokuma, the "Share" button is only available to Project Admins, right? So if I use the popover to downgrade myself from Admin to Viewer, I'll be ejected from the popover I think. Open to anything, but thoughts on this?

Ah yeah that's true. I guess to safeguard, I'm fine blocking that for now!

**Comment by ericpgreen2** [comment]
> > > > * I don't see the ability to set the default role for "epg" Organization Members. The ability to set this role should draw from the `manage_org_members` organization permission.
> > > 
> > > 
> > > I didn't implement this right off the bat because I wanted to keep this PR contained to project user management changeset. If we choose to include this org user role change from this project popover, we would have to individually set the role for each org user using `createAdminServiceSetOrganizationMemberUserRole`
> > > ```ts
> > > {
> > >   organization: string;
> > >   email: string;
> > >   data: {
> > >     role?: string;
> > >   };
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > and if we have [300 users](https://www.figma.com/design/Qt6EyotCBS3V6O31jVhMQ7/RILL-Latest?node-id=13326-670655&node-type=frame&t=cF5PVA45aU9jnfNk-0) in this project, we'd have to fire the request 300 times.
> > 
> > 
> > Oh I see. I agree that e.g. 300 individual requests would not be the way forward. We shouldn't be setting `user_orgs_roles` from this Project popover.
> > The designs seem to imply that we have an `organizations_projects_roles` table. We don't, but we do have a `usergroups_projects_roles` table. I could imagine an automatic `<organization>-users` usergroup for every organization, or I could imagine a dedicated `organizations_projects_roles` table. @begelundmuller @ericokuma, thoughts on this?
> 
> Ah, the designs were just assuming that all users who have access to an org will have access to projects. I think have a dedicated organizations_projects_roles table might be nice although wouldn't the all-users user group suffice?

Oh yeah, the `all-users` group is exactly the automatic `<organization>-users` usergroup I was thinking of! The "Viewer" toggle in the design implies that you can change Organization Members' default Project role from "Viewer" to "Admin", if you'd like. So to implement this @lovincyrus, the toggle should change the "all-users" Project role to "Admin".

image

**Comment by lovincyrus** [comment]
> > > > > * I don't see the ability to set the default role for "epg" Organization Members. The ability to set this role should draw from the `manage_org_members` organization permission.
> > > > 
> > > > 
> > > > I didn't implement this right off the bat because I wanted to keep this PR contained to project user management changeset. If we choose to include this org user role change from this project popover, we would have to individually set the role for each org user using `createAdminServiceSetOrganizationMemberUserRole`
> > > > ```ts
> > > > {
> > > >   organization: string;
> > > >   email: string;
> > > >   data: {
> > > >     role?: string;
> > > >   };
> > > > }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > and if we have [300 users](https://www.figma.com/design/Qt6EyotCBS3V6O31jVhMQ7/RILL-Latest?node-id=13326-670655&node-type=frame&t=cF5PVA45aU9jnfNk-0) in this project, we'd have to fire the request 300 times.
> > > 
> > > 
> > > Oh I see. I agree that e.g. 300 individual requests would not be the way forward. We shouldn't be setting `user_orgs_roles` from this Project popover.
> > > The designs seem to imply that we have an `organizations_projects_roles` table. We don't, but we do have a `usergroups_projects_roles` table. I could imagine an automatic `<organization>-users` usergroup for every organization, or I could imagine a dedicated `organizations_projects_roles` table. @begelundmuller @ericokuma, thoughts on this?
> > 
> > 
> > Ah, the designs were just assuming that all users who have access to an org will have access to projects. I think have a dedicated organizations_projects_roles table might be nice although wouldn't the all-users user group suffice?
> 
> Oh yeah, the `all-users` group is exactly the automatic `<organization>-users` usergroup I was thinking of! The "Viewer" toggle in the design implies that you can change Organization Members' default Project role from "Viewer" to "Admin", if you'd like. So to implement this @lovincyrus, the toggle should change the "all-users" Project role to "Admin".
> 
> image

`all-users` is a special usergroup that cannot be modified.

Look
CleanShot 2024-10-15 at 15 19 40@2x

**Comment by lovincyrus** [comment]
Since `all-users` at the project level is modifiable, should we keep the `all-users` group hidden from the Groups section? @ericokuma @ericpgreen2 

https://github.com/rilldata/rill/pull/5887#issuecomment-2414617888

File: web-admin/src/features/projects/user-invite/UserInviteOrganization.svelte (hunk 1 above)
[the remark this answers is not in the document]
Comment (lovincyrus) [reply web-admin/src/features/projects/user-invite/UserInviteOrganization.svelte:51]: > Which is the "project users table" you're referring to? In the mocks, I see users tables for Orgs & for User Groups, but not for Projects. But maybe I'm missing it or we have different terminology :)

Sorry, there is no "project users table". I was referring to the unified user management table.

[the remark this answers is not in the document]
Comment (ericpgreen2) [reply web-admin/src/features/projects/user-invite/UserInviteOrganization.svelte:51]: Got it, got it. So should we add a link to it? CC @jkhwu @ericokuma

[the remark this answers is not in the document]
Comment (ericpgreen2) [reply web-admin/src/features/projects/user-invite/UserInviteOrganization.svelte]: I discussed with Janet & we found that Notion does _not_ provide a link. Perhaps b/c it'd be too disruptive an act for a lightweight popover to change the app page.

File: web-admin/src/features/projects/user-management/ShareProjectPopover.svelte [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      {#if showOrganizationSection}
+        <div class="mt-4">
+          <div class="text-xs text-gray-500 font-semibold uppercase">
+            Organization
+          </div>
+          <div class="flex flex-col gap-y-1">
+            <UserInviteOrganization
+              {organization}
+              {project}
+              group={projectMemberUserGroupsList[0]}
```
Comment (ericpgreen2) [inline web-admin/src/features/projects/user-management/ShareProjectPopover.svelte:93]: Here and above, it looks like we're making the assumption that the first user group is the "all-users" group. To be safe & more explicit, could we explicitly get the "all-users" group by finding where `groupName === "all-users"`

File: web-admin/src/features/projects/user-management/ShareProjectPopover.svelte [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    createAdminServiceListProjectInvites,
+    createAdminServiceListProjectMemberUsergroups,
+    createAdminServiceListProjectMemberUsers,
+  } from "@rilldata/web-admin/client";
+  import CopyInviteLinkButton from "@rilldata/web-admin/features/projects/user-invite/CopyInviteLinkButton.svelte";
+  import UserInviteForm from "@rilldata/web-admin/features/projects/user-management/UserInviteForm.svelte";
+  import { Button } from "@rilldata/web-common/components/button";
+  import type { V1UserInvite } from "@rilldata/web-admin/client";
+  import UserInviteOrganization from "./UserManagementOrganization.svelte";
+  import UserInviteUserSetRole from "./UserManagementUserSetRole.svelte";
```
Comment (ericpgreen2) [inline web-admin/src/features/projects/user-management/ShareProjectPopover.svelte:14]: Still seeing the "UserInvite" terminology in the imports

File: web-admin/src/features/projects/user-invite/CopyInviteLinkButton.svelte
Comment (ericpgreen2) [inline web-admin/src/features/projects/user-invite/CopyInviteLinkButton.svelte:1]: Purposefully keeping this "user-invite" directory, or did you mean to move this to the "user-management" directory?

**Review by ericpgreen2 — COMMENTED** [review]
Last comment on component names: it's a style thing, but IMO it's redundant to add the "UserManagement" prefix to the components within the `user-management` directory. I would see more of a case for it if these components were used _outside_ the `user-management` directory. But in this case, all these "UserManagementX" components are getting imported into `.../features/projects/user-management/ShareProjectPopover.svelte`, so I don't think the extra prefix buys us anything.

File: web-admin/src/features/projects/user-management/ShareProjectPopover.svelte (hunk 2 above)
Comment (lovincyrus) [reply web-admin/src/features/projects/user-management/ShareProjectPopover.svelte:92]: There is already a check above for the Organization section.

```
  $: showOrganizationSection =
    projectMemberUserGroupsList.length === 1 &&
    projectMemberUserGroupsList[0].groupName === "all-users";
```

Comment (ericpgreen2) [reply web-admin/src/features/projects/user-management/ShareProjectPopover.svelte:90]: Yeah, so I think this makes two unsafe assumptions:
- only 1 group is returned from the API
- "all-users" is the first group in the list

What if there are >1 groups? We should still show the `OrganizationItem`, right?

I haven't inspected the API logic, but let's just say there's another group named "aaa". If the API ever returns ["aaa", "all-users"], then `showOrganizationSelection` should still be `true`.

**Review by ericpgreen2 — APPROVED** [review]
Looks clean! See my last reply before merging

**Pull request merged**